core/src/main/scala/kafka/server/LogOffsetMetadata.scala

```diff
@@ -58,8 +58,18 @@ case class LogOffsetMetadata(messageOffset: Long,
     this.segmentBaseOffset == that.segmentBaseOffset
   }
 
-  // check if this offset is before the given offset
-  def precedes(that: LogOffsetMetadata): Boolean = this.messageOffset < that.messageOffset
+  // check if this offset is before the given offset. We need to compare both message offset and segment base offset.
+  def precedes(that: LogOffsetMetadata): Boolean = {
+    this.messageOffset < that.messageOffset || {
+      if (this.messageOffsetOnly() && that.messageOffsetOnly())
+        false
+      else if (!this.messageOffsetOnly() && !that.messageOffsetOnly())
+        this.segmentBaseOffset < that.segmentBaseOffset
+      else
+        throw new IllegalArgumentException("Cannot compare " + this + " with " + that)
+    }
```

Comment:
Thanks for addressing my comment @becketqin. The other methods throw KafkaException, would it make sense to do the same here? Also, maybe the message could be a bit more informative like the other ones in the same class. Example:

```scala
throw new KafkaException(s"Cannot compare $this with $that as one has message offsets and the other does not.")
```